rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,454 @@
+  bots since it'll rapidly reschedule any bot deployments that fail... but we
+  have a dedicated `kubernetes` delegated join method.)
+
+End users willing to create their own automation around token issuance could
+work around some of these limitations, but this creates an unnecessary barrier
+to entry for use of Machine ID on-prem.
+
+## Details
+
+### Context: Priorities and Security Invariants
+
+It's become clear through conversations with customers and previous attempts at
+solving this problem that some of these pain points are contradictory. As a case
+study, an ideal UX from an end user's perspective might be to create one token,
+join all their bots with it, and leave them running indefinitely. This would
+create several issues:
+
+- The initial joining secret would have an ambiguous lifetime. At what point
+  does this multi-use token expire?
+
+- How many bots will join?
+
+- How can we tell joined bots apart? Can we trust a bot identity if it can be
+  thrown away and regenerated?
+
+- When a bot needs to rejoin, does it use the same token? Can that token *ever*
+  expire?
+
+With this in mind, we need to strike some balance between effective UX and a
+system we can trust to not allow unauthorized or unintended access. To that end,
+we'll focus on these explicit compromises we believe improve today's UX while
+making minimal security concessions:
+
+- **There must be a 1:1 relationship between a `ProvisionToken` and a bot
+  instance.** Allowing many joins on one token creates unnecessary backend
+  contention and - depending on implementation - creates severe traceability
+  problems.
+
+  However, we can greatly improve today's automation story. Secret fulfillment
+  can take place server side to reduce the number of resources to generate in
+  Terraform, and `ProvisionToken` resources can be made reusable to fully enable
+  IaC workflows.
+
+- **Bot credentials can be long-lived, but their *trust* must be controlled and
+  renewed.** We can't allow any secret values to either remain both valid and
+  useful for a long time, or allow currently valid identities to generate
+  credentials that can be used for future, unchecked extension of access.
+
+  However, we can create a new state where an identity is technically valid
+  indefinitely, but useless unless explicitly allowed by the server. Existing
+  controls like the generation counter can still effectively prevent unintended
+  reuse of the bot identity.
+
+### Challenge-Response Joining
+
+TODO: Consider alternative join method names?
+
+We believe a new join method, `challenge`, can meet our needs and provide
+significantly more flexibility than today's `token` join method. This works by -
+in a sense - inverting the token joining procedure: bots generate an ED25519
+keypair, and the public key is copied to the server. The public key can be
+copied out-of-band, or bots can provide their public key on first join using a
```

Comment:
I feel like I'm not understanding this correctly. If they can offer their public key on first join, what's to stop a malicious actor from just joining a bunch of bots? The security of the token is that the server generates it and knows it's valid. I get adding the public key out-of-band, but not this.

Ok, reading again, in this scenario where the public key is not added ahead of time, there would be an initial join secret, so basically a token would still be needed, but after that initial contact it wouldn't need the token again?

Reply:
For bulk joining, you'd want to use the Terraform provider with scenario 2, and have it generate an initial join secret for each node - there's an example of this further down in the document. The generated secret value can be provided to `tbot` and it will behave similarly to token joining today; keypair generation and exchange with Auth will be entirely transparent to the user.

Overall, in its default mode (i.e. scenario 2) it's functionally identical to traditional `token` joining for end users doing deployment, with a few enhancements to make Terraform automation a bit nicer (server side secret fulfillment, a sane story for restoring broken bots, etc).